optimizely/helpers/condition_tree_evaluator.py

```diff
@@ -0,0 +1,124 @@
+
+
+def and_evaluator(conditions, leaf_evaluator):
+  """ Evaluates a list of conditions as if the evaluator had been applied
+  to each entry and the results AND-ed together
+
+  Args:
+    conditions: List of conditions ex: [operand_1, operand_2]
+    leaf_evaluator: Function which will be called to evaluate leaf condition values
+
+  Returns:
+    Boolean:
+      - True if all operands evaluate to True
+      - False if a single operand evaluates to False
+    None: if conditions couldn't be evaluated
+  """
+  saw_null_result = False
+
+  for condition in conditions:
+    result = evaluate(condition, leaf_evaluator)
+    if result is False:
+      return False
+    if result is None:
+      saw_null_result = True
+
+  return None if saw_null_result else True
+
+
+def or_evaluator(conditions, leaf_evaluator):
+  """ Evaluates a list of conditions as if the evaluator had been applied
+  to each entry and the results OR-ed together
+
+  Args:
+    conditions: List of conditions ex: [operand_1, operand_2]
+    leaf_evaluator: Function which will be called to evaluate leaf condition values
+
+  Returns:
+    Boolean:
+      - True if any operand evaluates to True
+      - False if all operands evaluate to False
+    None: if conditions couldn't be evaluated
+  """
+  saw_null_result = False
+
+  for condition in conditions:
+    result = evaluate(condition, leaf_evaluator)
+    if result is True:
+      return True
+    if result is None:
+      saw_null_result = True
+
+  return None if saw_null_result else False
+
+
+def not_evaluator(conditions, leaf_evaluator):
+  """ Evaluates a list of conditions as if the evaluator had been applied
+  to a single entry and NOT was applied to the result.
+
+  Args:
+    conditions: List of conditions ex: [operand_1, operand_2]
+    leaf_evaluator: Function which will be called to evaluate leaf condition values
+
+  Returns:
+    Boolean:
+      - True if the operand evaluates to False
+      - False if the operand evaluates to True
+    None: if conditions is empty or condition couldn't be evaluated
+  """
+  if not len(conditions) > 0:
+    return None
+
+  result = evaluate(conditions[0], leaf_evaluator)
+  return None if result is None else not result
+
+EVALUATORS_BY_OPERATOR_TYPE = {
+  ConditionOperatorTypes.AND: and_evaluator,
+  ConditionOperatorTypes.OR: or_evaluator,
+  ConditionOperatorTypes.NOT: not_evaluator
+}
+
+
+def evaluate(conditions, leaf_evaluator):
+  """ Top level method to evaluate conditions
+
+  Args:
+    conditions: Nested array of and/or conditions, or a single leaf condition value of any type
+                Example: ['and', '0', ['or', '1', '2']]
+    leaf_evaluator: Function which will be called to evaluate leaf condition values
+
+  Returns:
+    Boolean: Result of evaluating the conditions using the operator rules and the leaf evaluator.
+    None: if conditions couldn't be evaluated
+
+  """
+
+  if isinstance(conditions, list):
+    if conditions[0] in OPERATOR_TYPES:
```

Comment:
@nchilada Can't use EVALUATORS_BY_OPERATOR_TYPE here as in implicit OR case the first item in conditions[0] is a dict which is an unhashable type. Using EVALUATORS_BY_OPERATOR_TYPE.items() has different result types for python 2 and 3

Reply:
> Using EVALUATORS_BY_OPERATOR_TYPE.keys() has different result types for python 2 and 3

I'm actually not very familiar with Python 3, but could we try
```python
if conditions[0] in EVALUATORS_BY_OPERATOR_TYPE:
```
? Since that's the equivalent of
```python
if conditions[0] in EVALUATORS_BY_OPERATOR_TYPE.keys():
```
in Python 2